Add GetTabRect() and GetTabOrientation() to wxNotebook.

skeetor opened the pull request:
Add GetTabOrientation() to return the orientation of the tabs on a wxNotebook.
Add GetTabRect() to return the window coordinates of a tab from a wxNotebook.

skeetor committed:
Add GetTabRect() to return the position of a tab from a wxNotebook.
Add GetTabOrientation() to return the orientation of the tabs on a wxNotebook.
Documentation updated
Added testcases

PBfordev commented on include/wx/msw/notebook.h:
```diff
@@ -93,6 +93,9 @@ class WXDLLIMPEXP_CORE wxNotebook : public wxNotebookBase
     // style.
   void SetTabSize(const wxSize& sz) override;
 
+  // Return the position and size of he tab for the given page
```
```suggestion
  // return the position and size of the tab for the given page
```

skeetor replied:
Fixed

PBfordev commented on include/wx/notebook.h:
```diff
@@ -139,6 +139,13 @@ class WXDLLIMPEXP_CORE wxNotebookBase : public wxBookCtrlBase
+    // return wxTOP/wxBOTTOM/wxRIGHT/wxLEFT
+    wxDirection GetTabOrientation() const;
+
+    // return the rectangle in windowcoordinates of the page selector.
```
```suggestion
    // return the rectangle in window coordinates of the page selector
```

skeetor replied:
Fixed

PBfordev commented on include/wx/notebook.h:
```diff
@@ -139,6 +139,13 @@ class WXDLLIMPEXP_CORE wxNotebookBase : public wxBookCtrlBase
+    wxDirection GetTabOrientation() const;
+
+    // return the rectangle in windowcoordinates of the page selector.
+    // i.E. for a wxNotebook this would be it's tab.
```
```suggestion
    // e.g. for a wxNotebook this would be its tab
```

skeetor replied:
Fixed

PBfordev commented on include/wx/notebook.h:
```diff
@@ -139,6 +139,13 @@ class WXDLLIMPEXP_CORE wxNotebookBase : public wxBookCtrlBase
+
+    // return the rectangle in windowcoordinates of the page selector.
+    // i.E. for a wxNotebook this would be it's tab.
+    virtual wxRect GetTabRect(size_t n) const;
```
```suggestion
    virtual wxRect GetTabRect(size_t page) const;
```

skeetor replied:
Fixed

PBfordev commented on interface/wx/notebook.h:
```diff
@@ -191,6 +191,19 @@ class wxNotebook : public wxBookCtrlBase
+    wxDirection GetTabOrientation() const;
+
+    /**
+        Return the rectangle in windowcoordinates of the page selector.
```
```suggestion
        Returns the rectangle in window coordinates of the page selector.
```

skeetor replied:
Fixed

PBfordev commented on interface/wx/notebook.h:
```diff
@@ -191,6 +191,19 @@ class wxNotebook : public wxBookCtrlBase
+
+    /**
+        Return the rectangle in windowcoordinates of the page selector.
+        i.E. for a wxNotebook this would be it's tab.
```
```suggestion
        For example, for a wxNotebook this would be its tab.
```

skeetor replied:
Fixed

PBfordev commented on interface/wx/notebook.h:
```diff
@@ -191,6 +191,19 @@ class wxNotebook : public wxBookCtrlBase
+
+        @note Currently only available in Univ and MSW
+    */
+    virtual wxRect GetTabRect(size_t n) const;
```
```suggestion
    virtual wxRect GetTabRect(size_t page) const;
```

skeetor replied:
Fixed

PBfordev commented on src/common/nbkbase.cpp:
```diff
@@ -197,4 +197,25 @@ void wxNotebookBase::SendPageChangedEvent(int nPageOld, int nPageNew)
+
+wxRect wxNotebookBase::GetTabRect(size_t WXUNUSED(n)) const
+{
+    wxFAIL_MSG(wxT("wxNotebook::GetTabRect not implemented"));
```
```suggestion
    wxFAIL_MSG("not implemented");
```

skeetor replied:
Fixed

PBfordev commented on src/msw/notebook.cpp:
```diff
@@ -499,6 +499,25 @@ void wxNotebook::SetTabSize(const wxSize& sz)
+        RECT rect;
+        if (TabCtrl_GetItemRect(GetHwnd(), n, &rect))
+        {
+            r.x = rect.left;
```
I would use `r = wxRectFromRECT(rect);`  instead of all these four lines.

skeetor replied:
Fixed. Didn't know about this funtion. :)

PBfordev commented on src/univ/notebook.cpp:
```diff
@@ -688,7 +674,7 @@ wxRect wxNotebook::GetTabRect(int page) const
     {
         widthBefore = 0;
-        for ( int n = 0; n < page; n++ )
+        for (size_t n = 0; n < page; n++)
```
```suggestion
        for ( size_t n = 0; n < page; n++ )
```

skeetor replied:
Already fixed in latest commit

vadz replied:
This is pretty minor, but it doesn't seem to have been fixed.

PBfordev reviewed (commented):
Code specific comments aside...

Not sure if the maintainers accept such methods when implemented only for one of the three major ports. I am MSW-only myself, so I have no idea if it is possible to do the same for macOS and GTK.

Just curious, does it work correctly when the tabs are not on the top (can be easily tested with the notebook sample)? IIRC, there were some issues on MSW with non-top tab placement, you can see in the sample that the tabs look different in other positions than the top.

skeetor commented:
This code definitely works on MSW and I tested it with all orientations of the tabs. For the univ build, the functions already existed and I didn't change them, so I assume they work or at least will behave as they always had.

I would like to add this to QT and GTK build later on as well, but currently I work on my main implementation of my module and I don't want to get diverted to much by it by testing several other builds as well. However, since these functions were not there before anyway, they don't brake anything as of now.

PBfordev commented on src/msw/notebook.cpp:
```diff
@@ -499,6 +499,20 @@ void wxNotebook::SetTabSize(const wxSize& sz)
+    {
+        RECT rect;
+        if (TabCtrl_GetItemRect(GetHwnd(), n, &rect))
+            wxRectFromRECT(rect);
```
```suggestion
            r = wxRectFromRECT(rect);
```

BTW, I meant to change the method argument name from `n` to `page` everywhere it was used (similarly for typos in the comments).

skeetor replied:
Fixed

oneeyeman1 commented:
@skeetor ,
There was a bug in Windows about HitTest not working correctly for non-top tab position. MS confirmed it.

I presume this code works as it should.

Maybe we can implement HitTest somehow with this change for non-top tab position?

skeetor commented:
I noticed that about HitTest because I tried using that first. Since it didn't work as I expected it, I wrote some code to detect klicks in the off region of the tabs, why I needed this functions. :)
I wrote some code to get the full tab rectangle, as well as getting the rectangle of the off region. So should I add this as well, so it might be used by HitTest?

oneeyeman1 commented:
@skeetor ,
There is a long-withstanding issue/ticket about it.
I told MS about it and they confirmed it.
If you can make HitTest() on MSW work as expected - it would be great.
However, it probably should be another PR and if you can find this issue and reference it it will be even better.

skeetor commented:
OK! I would also prefer to have it in another PR anyway so this would be fine for me.

skeetor commented:
The only reference I could find is this one: https://github.com/wxWidgets/wxWidgets/issues/4775

However, with my (upcoming) fix this will work for all orientations, as I needed this to work reliably for my docking module.

vadz commented on interface/wx/notebook.h:
```diff
@@ -191,6 +191,19 @@ class wxNotebook : public wxBookCtrlBase
     virtual void SetPadding(const wxSize& padding);
 
+    /**
+        Returns wxTOP / wxBOTTOM / wxRIGHT / wxLEFT
```
I'd like to expand the description a bit and we also should indicate that it's only available in 3.3 and later:

```suggestion
        Returns one of wxTOP, wxBOTTOM, wxRIGHT or wxLEFT.
        
        This is a convenience function mapping wxBK_TOP etc styles to one of the
        wxDirection enum elements.
        
        @since 3.3.0
```

vadz commented on interface/wx/notebook.h:
```diff
@@ -191,6 +191,19 @@ class wxNotebook : public wxBookCtrlBase
+        For example, for a wxNotebook this would be its tab.
+
+        @note Currently only available in Univ and MSW
+    */
```
```suggestion
        @since 3.3.0
    */
```

vadz commented on interface/wx/notebook.h:
```diff
@@ -191,6 +191,19 @@ class wxNotebook : public wxBookCtrlBase
+
+    /**
+        Return the rectangle in window coordinates of the page selector.
+        For example, for a wxNotebook this would be its tab.
```
I'm not sure it makes sense to say "for a wxNotebook" in wxNotebook documentation. Was this supposed to be a `wxBookCtrlBase` function initially?

vadz commented on include/wx/notebook.h:
```diff
@@ -139,6 +139,13 @@ class WXDLLIMPEXP_CORE wxNotebookBase : public wxBookCtrlBase
+    // return wxTOP/wxBOTTOM/wxRIGHT/wxLEFT
+    wxDirection GetTabOrientation() const;
+
+    // return the rectangle in window coordinates of the page selector.
+    // For example, for a wxNotebook this would be its tab.
```
As also mentioned in the interface file comment, saying "for example for a wxNotebook" in wxNotebook documentation doesn't seem to make much sense, so I'd just write
```suggestion
    // return the rectangle of the given page tab in window coordinates
```

vadz commented on src/common/nbkbase.cpp:
```diff
@@ -197,4 +197,25 @@ void wxNotebookBase::SendPageChangedEvent(int nPageOld, int nPageNew)
+
+wxRect wxNotebookBase::GetTabRect(size_t WXUNUSED(n)) const
+{
+    wxFAIL_MSG(wxT("Not implemented"));
```
Let's not use `wxT()` in the new code:
```suggestion
    wxFAIL_MSG("Not implemented");
```

vadz reviewed (commented):
Thanks, I'm not really against having this function even if it's MSW-only if it can be implemented for the other ports later, but I have some questions about it that should be answered in the documentation and, ideally, tests:

1. What happens when the tab is scrolled away, i.e. not currently visible at all?
2. What happens if an invalid index (too big) is passed?
3. Do we really want to assert in the other ports or maybe we should just return an empty rectangle silently? I think the latter would be more useful as otherwise I don't really know how would I use this function in portable code.

Having a test for this function in `tests/controls/notebooktest.cpp` would be really, really nice -- it could be used to clarify the answers to the questions above and also make it simpler to check that the function is implemented for the other platforms correctly later.

TIA!

skeetor commented:
>     1. What happens when the tab is scrolled away, i.e. not currently visible at all?

The official MSW function TabCtrl_GetItemRect() reports the position where the tab would be. There is no indicator that the tab is not visible. So the application has to check this itself if it loops through all tabs.

>     2. What happens if an invalid index (too big) is passed?

I added the checks.

>     3. Do we really want to assert in the other ports or maybe we should just return an empty rectangle silently? I think the latter would be more useful as otherwise I don't really know how would I use this function in portable code.

Personally I prefer to get a big red warning from the library layer instead of silently giving me a wrong answer. If an application developer wants to use this function and it is not implemented on his platform, then he should be aware that it is not working, instead of having to debug it why it doesn't work only to find out that it's not implemented.
I was considering to implement it later on for wxGTK myself, when I get to port my app to linux, but this is still far in the future, so maybe somebody may beat me to it. :) As for other platforms like OSX, somebody, who is interested enough would have to do it.

> Having a test for this function in `tests/controls/notebooktest.cpp` would be really, really nice -- it could be used to clarify the answers to the questions above and also make it simpler to check that the function is implemented for the other platforms correctly later.

OK. I will take a look at it how to test it. I added all your other suggestions as well.

vadz commented:
> > 1. What happens when the tab is scrolled away, i.e. not currently visible at all?
> 
> The official MSW function TabCtrl_GetItemRect() reports the position where the tab would be. There is no indicator that the tab is not visible. So the application has to check this itself if it loops through all tabs.

This is fine, but please add this to the documentation, as it's not really obvious. I.e. just say that the position of the tab can be beyond the window boundary or something like this.

> > 2. What happens if an invalid index (too big) is passed?
> 
> I added the checks.

Thanks!

> > 3. Do we really want to assert in the other ports or maybe we should just return an empty rectangle silently? I think the latter would be more useful as otherwise I don't really know how would I use this function in portable code.
> 
> Personally I prefer to get a big red warning from the library layer instead of silently giving me a wrong answer. If an application developer wants to use this function and it is not implemented on his platform, then he should be aware that it is not working, instead of having to debug it why it doesn't work only to find out that it's not implemented.

This is a reasonable position, although I'm still not sure if it's the best thing to do in this particular case. But we can start by doing this as we can always remove the assert later if it's really problematic. However it should, again, be clearly documented, i.e. the docs should say that this function is only implemented in wxMSW and wxUniv and asserts if it's called under another platform.

skeetor commented:
> Having a test for this function in `tests/controls/notebooktest.cpp` would be really, really nice -- it could be used to clarify the answers to the questions above and also make it simpler to check that the function is implemented for the other platforms correctly later.

I'm currently doing the test, but that got me thinking. When I add a testcase for GetTabRect() wont the builds for GTK and OSX fail because of the assert, until these functions are properly implemented?

vadz commented:
> > Having a test for this function in `tests/controls/notebooktest.cpp` would be really, really nice -- it could be used to clarify the answers to the questions above and also make it simpler to check that the function is implemented for the other platforms correctly later.
> 
> I'm currently doing the test, but that got me thinking. When I add a testcase for GetTabRect() wont the builds for GTK and OSX fail because of the assert, until these functions are properly implemented?

Yes, you need to use `#if defined(__WXMSW__) || defined(__WXUNIVERSAL__)` around these tests for now.

vadz commented:
Thanks, applied with some minor/cosmetic changes.
